### PR TITLE
CAMEL-15622: Prefer bean instance in endpoint DSL

### PR DESCRIPTION
**Pull request by Drezir**
Do not search in registry using wrong string representation of a bean, return the bean directly instead

- [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/CAMEL) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
- [x] Each commit in the pull request should have a meaningful subject line and body.
- [x] If you're unsure, you can format the pull request title like `[CAMEL-XXX] Fixes bug in camel-file component`, where you replace `CAMEL-XXX` with the appropriate JIRA issue.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Run `mvn clean install -Psourcecheck` in your module with source check enabled to make sure basic checks pass and there are no checkstyle violations. A more thorough check will be performed on your pull request automatically.
Below are the contribution guidelines:
https://github.com/apache/camel/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: core/camel-support/src/main/java/org/apache/camel/support/DefaultComponent.java [hunk 1]
```diff
@@ -551,6 +551,9 @@ protected boolean useIntrospectionOnEndpoint() {
      * @throws IllegalArgumentException if referenced object was not found in registry.
      */
     public <T> T resolveAndRemoveReferenceParameter(Map<String, Object> parameters, String key, Class<T> type, T defaultValue) {
+        if (parameters.containsKey(key) && type.isInstance(parameters.get(key).getClass())) {
```
Comment (bedlaj) [inline core/camel-support/src/main/java/org/apache/camel/support/DefaultComponent.java:554]: Parameter value can be null. Actually many tests are failing with `NullPointerException`.

**Review by bedlaj — CHANGES_REQUESTED** [review]
Hi, thanks for contribution. Unfortunately your change fails with NPE in many situations.

File: core/camel-support/src/main/java/org/apache/camel/support/DefaultComponent.java (hunk 1 above)
Comment (Drezir) [reply core/camel-support/src/main/java/org/apache/camel/support/DefaultComponent.java]: Should be OK now. Thanks.

Comment (bedlaj) [reply core/camel-support/src/main/java/org/apache/camel/support/DefaultComponent.java]: No. `parameters` will be never null. But value in map (`parameters.get(key)`) can be null. The NPE is still there. You can rollback last commit and instead replace `parameters.containsKey(key)` with `parameters.get(key) != null`

**Review by bedlaj — CHANGES_REQUESTED** [review]
NPE is still there, as value in `parameters` map can be null.

File: core/camel-support/src/main/java/org/apache/camel/support/DefaultComponent.java (hunk 1 above)
Comment (Drezir) [reply core/camel-support/src/main/java/org/apache/camel/support/DefaultComponent.java]: Done, thanks

**Review by bedlaj — CHANGES_REQUESTED** [review]
I have still some doubts.

- `isInstance` method expects instance object, but you are passing `Class`. Your condition will be always false. Can you fix that and also please add unit test to see if this PR actually fixes the the issue, as currently it does not, your code is never hitted.

- What hapens with String beans? Eg something like this will probably not work once you fix the previous point. It will just return `#myJmsUsername` without resolving that to `brokerUserName` String.

```
@BindToRegistry("myJmsUsername")
String myJmsUsername = "brokerUserName";

from(jms(...).username("#myJmsUsername"));

```

File: core/camel-support/src/main/java/org/apache/camel/support/DefaultComponent.java [hunk 2]
```diff
@@ -551,6 +551,9 @@ protected boolean useIntrospectionOnEndpoint() {
      * @throws IllegalArgumentException if referenced object was not found in registry.
      */
     public <T> T resolveAndRemoveReferenceParameter(Map<String, Object> parameters, String key, Class<T> type, T defaultValue) {
+        if (parameters.get(key) != null && type.isInstance(parameters.get(key))) {
```
Comment (bedlaj) [inline core/camel-support/src/main/java/org/apache/camel/support/DefaultComponent.java:554]: Now, when is `getClass` removed, the null check is redundant, as `type.isInstance(null)` returns false. The condition can be simplified to `if (type.isInstance(parameters.get(key)))`

**Comment by davsclaus** [comment]
> I have still some doubts.
> 
> * `isInstance` method expects instance object, but you are passing `Class`. Your condition will be always false. Can you fix that and also please add unit test to see if this PR actually fixes the the issue, as currently it does not, your code is never hitted.
> * What hapens with String beans? Eg something like this will probably not work once you fix the previous point. It will just return `#myJmsUsername` without resolving that to `brokerUserName` String.
> 
> ```
> @BindToRegistry("myJmsUsername")
> String myJmsUsername = "brokerUserName";
> 
> from(jms(...).username("#myJmsUsername"));
> ```

The endpointdsl works differently for those simple types. This problem is for complex object types that are set as additional parameters form the uri (special for endpoint dsl).

So this PR is in the right direction.

**Comment by davsclaus** [comment]
I have merged this PR manually and added some tests and improved the code in other parts where it was needed. Doing some tests before merging.

**Comment by bedlaj** [comment]
Thanks again for contribution @Drezir . Are you registered in hacktoberfest? Since Claus will merge this PR manually, the PR will be closed and you would not get the PR counted to score. Let us know, if you are competing in hacktoberfest. This PR was helpful, and we can mark the PR as `hacktoberfest-accepted` to give you credit even for closed PR.

**Comment by davsclaus** [comment]
The PR has one of @Drezir contributions here
https://github.com/apache/camel/commit/2b13a0138e42b4f27623cb637845d970c907ad96

so that ought to get a score for you.

**Pull request closed without merging**